### PR TITLE
Add SDPA and FlashAttention-2 support to LayoutLMv3

### LINKED ISSUE #35467: Support Flash Attention 2 for LayoutLMv3
### Feature request

Support Flash Attention 2 for LayoutLMv3.

### Motivation

Speed up document transformer training and inference.

### Your contribution

I can add it.

### PR DESCRIPTION
**Pull request by jackiehimel**
# What does this PR do?

Adds SDPA and FlashAttention-2 support to LayoutLMv3 using the unified attention interface pattern, following the same architecture used in BERT and other recent model implementations.

Fixes #35467

## Implementation

- Refactored `LayoutLMv3SelfAttention` to use `ALL_ATTENTION_FUNCTIONS` interface instead of separate attention classes
- Created `layoutlmv3_eager_attention_forward` function that implements the CogView attention mechanism (alpha=32 scaling) with support for LayoutLMv3's relative position bias and spatial attention bias
- Added `_supports_sdpa = True` and `_supports_flash_attn = True` flags to `LayoutLMv3PreTrainedModel`
- Updated mask creation to use `create_bidirectional_mask` (replacing `get_extended_attention_mask`)
- Threaded `layer_idx` parameter through attention classes for consistency
- **Added automatic enforcement in `LayoutLMv3Config`** to set `attn_implementation="eager"` when relative or spatial attention biases are enabled (default behavior)

**Note**: SDPA and FlashAttention-2 are incompatible with LayoutLMv3's relative position bias and spatial attention bias. The config automatically enforces eager attention when these biases are enabled (the default). To use SDPA/FlashAttention-2, users must disable both biases (has_relative_attention_bias=False and has_spatial_attention_bias=False).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How has this change been tested?

- CI tests passing, linting and formatting locally validated.
- Added test skips for SDPA/Flash comparison tests (LayoutLMv3 defaults to eager when biases are enabled) and overrode `test_batching_equivalence` to ensure eager attention is used.
- Implementation follows the unified attention pattern from BERT.

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

@vasqu - Thanks for the feedback on #41801! I've refactored this to use the unified attention interface pattern as you suggested. Would appreciate another look when you have time.

@ArthurZucker @CyrilVallez - attention implementation reviewers

### CODE REVIEW

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 1]
```diff
@@ -214,18 +220,80 @@ def _init_weights(self, module):
...
+    LayoutLMv3 eager attention with support for relative position bias and spatial attention bias.
+    Based on the CogView attention trick for training stability.
+    """
+    if scaling is None:
+        scaling = 1.0 / math.sqrt(query.size(-1))
+
+    # Take the dot product between "query" and "key" to get the raw attention scores.
+    # The attention scores QT K/√d could be significantly larger than input elements, and result in overflow.
+    # Changing the computational order into QT(K/√d) alleviates the problem. (https://huggingface.co/papers/2105.13290)
+    attention_scores = torch.matmul(query / math.sqrt(query.size(-1)), key.transpose(-1, -2))
```
Comment (vasqu) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py]: It's essentially the same as bert https://github.com/huggingface/transformers/blob/47227f477532e8ff6dd3290dc6b9e0e38aef303b/src/transformers/models/bert/modeling_bert.py#L131

with flipped order: `attn_weights = torch.matmul(query * scaling, key.transpose(2, 3))`

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 2]
```diff
@@ -214,18 +220,80 @@ def _init_weights(self, module):
...
+        module.has_relative_attention_bias
+        and module.has_spatial_attention_bias
+        and rel_pos is not None
+        and rel_2d_pos is not None
+    ):
+        attention_scores = attention_scores + (rel_pos + rel_2d_pos) / math.sqrt(query.size(-1))
+    elif module.has_relative_attention_bias and rel_pos is not None:
+        attention_scores = attention_scores + rel_pos / math.sqrt(query.size(-1))
+    elif module.has_spatial_attention_bias and rel_2d_pos is not None:
+        attention_scores = attention_scores + rel_2d_pos / math.sqrt(query.size(-1))
```
Comment (vasqu) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py]: Are we applying the scaling twice here? Do the integration tests still pass? I.e. another ` / math.sqrt(query.size(-1)`

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 3]
```diff
@@ -234,18 +302,7 @@ def __init__(self, config):
         self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
         self.has_relative_attention_bias = config.has_relative_attention_bias
         self.has_spatial_attention_bias = config.has_spatial_attention_bias
```
Comment (vasqu) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:240]: It's missing an `is_causal` attribute, I would think it's not causal

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 4]
```diff
@@ -254,54 +311,45 @@ def forward(
         output_attentions=False,
```
Comment (vasqu) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:258]:
```suggestion
```

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 5]
```diff
@@ -254,54 +311,45 @@ def forward(
...
+
+        if not use_eager:
+            # SDPA and Flash Attention don't support custom relative position bias and spatial attention bias
+            if self.has_relative_attention_bias or self.has_spatial_attention_bias:
+                raise ValueError(
+                    f"You are using {self.config._attn_implementation} as attention type. However, LayoutLMv3's "
+                    "relative position bias and spatial attention bias are not compatible with it. "
+                    'Please load the model with `attn_implementation="eager"`.'
+                )
+            attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```
Comment (vasqu) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py]:
```suggestion
        if self.config._attn_implementation != "eager":
            # SDPA and Flash Attention don't support custom relative position bias and spatial attention bias
            if self.has_relative_attention_bias or self.has_spatial_attention_bias:
                raise ValueError(
                    f"You are using {self.config._attn_implementation} as attention type. However, LayoutLMv3's "
                    "relative position bias and spatial attention bias are not compatible with it. "
                    'Please load the model with `attn_implementation="eager"`.'
                )
            attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```
For now we raise a value error, thinking about forcing a fallback instead. Let's keep it as is for now

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 6]
```diff
@@ -254,54 +311,45 @@ def forward(
...
+            dropout=0.0 if not self.training else self.dropout.p,
+            scaling=self.scaling,
+            rel_pos=rel_pos,
+            rel_2d_pos=rel_2d_pos,
+            **kwargs,
+        )
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
 
+        outputs = (attn_output, attn_weights) if output_attentions else (attn_output,)
         return outputs
```
Comment (vasqu) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py]:
```suggestion
        return outputs, attn_weights
```
You can take a look at https://github.com/huggingface/transformers/blob/47227f477532e8ff6dd3290dc6b9e0e38aef303b/src/transformers/models/bert/modeling_bert.py#L560-L562

which takes care of the output xxx. Will need more changes here then but better to go for the right thing from the get go.

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 7]
```diff
@@ -320,11 +368,11 @@ def forward(self, hidden_states: torch.Tensor, input_tensor: torch.Tensor) -> to
         return hidden_states
 
 
-# Copied from transformers.models.layoutlmv2.modeling_layoutlmv2.LayoutLMv2Attention with LayoutLMv2->LayoutLMv3
+# Adapted from transformers.models.layoutlmv2.modeling_layoutlmv2.LayoutLMv2Attention with LayoutLMv2->LayoutLMv3
```
Comment (vasqu) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py]: Is layoutlm v2 vastly different, might be worth to change along at the same time.

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py (hunk 6 above)
Comment (vasqu) [reply src/transformers/models/layoutlmv3/modeling_layoutlmv3.py]: Bert is a good reference for this + you need the same decorators, they are essential.

File: tests/models/layoutlmv3/test_modeling_layoutlmv3.py [hunk 8]
```diff
@@ -162,12 +170,17 @@ def prepare_config_and_inputs(self):
...
+        # which is incompatible with SDPA/FlashAttention
+        config._attn_implementation = "eager"
 
         return config, input_ids, bbox, pixel_values, token_type_ids, input_mask, sequence_labels, token_labels
 
     def create_and_check_model(
         self, config, input_ids, bbox, pixel_values, token_type_ids, input_mask, sequence_labels, token_labels
     ):
+        # Ensure eager attention is set before model creation
+        config._attn_implementation = "eager"
```
Comment (vasqu) [inline tests/models/layoutlmv3/test_modeling_layoutlmv3.py]: Is the change in `prepare_config_and_inputs` not enough? Same for below

File: tests/models/layoutlmv3/test_modeling_layoutlmv3.py [hunk 9]
```diff
@@ -309,6 +331,55 @@ def setUp(self):
         self.model_tester = LayoutLMv3ModelTester(self)
         self.config_tester = ConfigTester(self, config_class=LayoutLMv3Config, hidden_size=37)
 
+    @parameterized.expand(TEST_EAGER_MATCHES_SDPA_INFERENCE_PARAMETERIZATION)
+    @unittest.skip("LayoutLMv3's relative position bias and spatial attention bias are incompatible with SDPA.")
+    def test_eager_matches_sdpa_inference(self, *args):
+        pass
```
Comment (vasqu) [inline tests/models/layoutlmv3/test_modeling_layoutlmv3.py]: Does that mean that layout lm always uses the relative bias? Then it might not even make sense to support sdpa/flash attn - it won't be used either way

**Review by vasqu — COMMENTED** [review]
Sorry, I did go through the code and gave some comments but I'm noticing that layout lm only uses the relative bias. If that's the case, then the usage of other attention flavors is questionable as they won't be used either way.

Imo, it would make more sense to go for other models that are suitable. You already did a overall pretty good job over here.
